### PR TITLE
Add a notebook for torchdata.nodes basics

### LINKED ISSUE #1352: [WIP] Examples for demonstrating the usage and incremental value of TorchData Nodes
### 🚀 The feature

Starting this issue to track minimal examples we can create to demonstrate effective usage and value of TorchData nodes. I can create separate issues for each of these as required.

### Motivation, pitch

1. Vanilla torch.utils dataloader usage ported over to torchdata nodes
2. GPU accelerated transforms
3. Flexible parallelism (mixing multiprocessing with multithreading)
4. Examples porting over popular OSS datasets
    -  connecting to popular cloud storage
5. Example creating new nodes (might get covered through examples above)
6. Basic multimodal model trained E2E using torchdata nodes
7. Chaining multiple transforms (might get covered through examples above)
8. Dataset mixing (with different sampling strategies)

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ramanishsingh**
Adding torchdata.notes basic examples
Fixes #1352

### CODE REVIEW

File: examples/nodes/torchata_nodes_basics.ipynb [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+   "source": [
+    "#### Let's see the functionalities of `torchdata.nodes` through the help of a very simple example"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "b5168248-7c08-419d-a9a5-f831bd7b46e7",
+   "metadata": {},
+   "source": [
+    "#### BaseNode"
```
Comment (andrewkho) [inline examples/nodes/torchata_nodes_basics.ipynb:45]: This should be IterableWrapper

File: examples/nodes/torchata_nodes_basics.ipynb [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+      "16\n",
+      "25\n",
+      "36\n",
+      "49\n",
+      "64\n",
+      "81\n"
+     ]
+    }
+   ],
+   "source": [
```
Comment (andrewkho) [inline examples/nodes/torchata_nodes_basics.ipynb:196]: Add a `ParallelMapper` heading here

File: examples/nodes/torchata_nodes_basics.ipynb [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+      "5\n",
+      "7\n",
+      "0\n",
+      "8\n",
+      "1\n",
+      "9\n"
+     ]
+    }
+   ],
+   "source": [
```
Comment (andrewkho) [inline examples/nodes/torchata_nodes_basics.ipynb:122]: Can we add a hyperlinnk to the (just landed) migration guide?

File: examples/nodes/torchata_nodes_basics.ipynb [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+   "execution_count": 1,
+   "id": "f42819f7-1019-48ca-ad23-00b38c72d63e",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from torchdata.nodes import IterableWrapper\n",
+    "# This Wrapper that converts any Iterable in to a BaseNode.\n",
+    "\n",
+    "dataset = range(10) # creating a very simple dataset, and then converting it into a BaseNode\n",
+    "base_node = IterableWrapper(dataset)"
```
Comment (andrewkho) [inline examples/nodes/torchata_nodes_basics.ipynb:59]: Let's call this `source`

File: examples/nodes/torchata_nodes_basics.ipynb [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+      "49\n",
+      "64\n",
+      "81\n"
+     ]
+    }
+   ],
+   "source": [
+    "# We can use the Mapper class, to apply a transformation defined using the `map_fn`\n",
+    "from torchdata.nodes import Mapper\n",
+    "mapped_dataset = Mapper(base_node, map_fn = lambda x : x**2)\n",
```
Comment (andrewkho) [inline examples/nodes/torchata_nodes_basics.ipynb:168]: Let's call this `node` too

File: examples/nodes/torchata_nodes_basics.ipynb [hunk 6]
```diff
@@ -0,0 +1,406 @@
...
+      "49\n",
+      "64\n",
+      "81\n"
+     ]
+    }
+   ],
+   "source": [
+    "# It can also be executed in parallel, using the multi threading/processing approaches, depending the defined `method`\n",
+    "from torchdata.nodes import ParallelMapper\n",
+    "mapped_dataset = ParallelMapper(base_node, map_fn = lambda x : x**2, num_workers =2, method = \"thread\")\n",
```
Comment (andrewkho) [inline examples/nodes/torchata_nodes_basics.ipynb:199]: ditto just call this `node`

File: examples/nodes/torchata_nodes_basics.ipynb (hunk 2 above)
Comment (andrewkho) [reply examples/nodes/torchata_nodes_basics.ipynb:196]: actually, nevermind this is OK

File: examples/nodes/torchata_nodes_basics.ipynb [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+      "[0, 1, 2, 3]\n",
+      "[4, 5, 6, 7]\n"
+     ]
+    }
+   ],
+   "source": [
+    "# A BaseNode can be passed into a Batcher, to get batches of size `batch_size`.\n",
+    "# By default, drop_last is True, meaning if the last batch has a size smaller than the `batch_size`, it is not produced.\n",
+    "from torchdata.nodes import Batcher\n",
+    "batched_dataset = Batcher(base_node, batch_size = 4)\n",
```
Comment (andrewkho) [inline examples/nodes/torchata_nodes_basics.ipynb:231]: just call this `batched`

File: examples/nodes/torchata_nodes_basics.ipynb [hunk 8]
```diff
@@ -0,0 +1,406 @@
...
+   "source": [
+    "# torchadata nodes are composable, thus, many BaseNodes type nodes can be chained together for desired transformations"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "5f1a6291-d688-45eb-bf2e-56153419513e",
+   "metadata": {},
+   "source": [
+    "#### Chaining multiple operations together"
```
Comment (andrewkho) [inline examples/nodes/torchata_nodes_basics.ipynb:346]:
```suggestion
    "#### Chaining multiple operations together\n",
    "Base nodes are iterators, and are designed to be chained together to create more complex dataloading graphs."
```

File: examples/nodes/torchata_nodes_basics.ipynb [hunk 9]
```diff
@@ -0,0 +1,406 @@
...
+   "execution_count": 11,
+   "id": "cc4bcd30-9a76-4a1e-aa38-d7db8b9a8c2e",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "sampler = RandomSampler(dataset)\n",
+    "node = MapStyleWrapper(map_dataset=dataset, sampler=sampler)\n",
+    "mapped_dataset = Mapper(node, map_fn = lambda x : x**3)\n",
+    "batched_dataset = Batcher(mapped_dataset, batch_size = 4, drop_last = False)\n",
+    "loader = Loader(batched_dataset)"
```
Comment (andrewkho) [inline examples/nodes/torchata_nodes_basics.ipynb:360]:
```suggestion
    "sampler = RandomSampler(dataset)\n",
    "node = MapStyleWrapper(map_dataset=dataset, sampler=sampler)\n",
    "node = Mapper(node, map_fn = lambda x : x**3)\n",
    "node = Batcher(node, batch_size = 4, drop_last = False)\n",
    "loader = Loader(node)"
```

File: examples/nodes/torchata_nodes_basics.ipynb [hunk 10]
```diff
@@ -0,0 +1,406 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "id": "e5384863-fa3e-450e-a8fc-e838e4d2f432",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# torchadata nodes are composable, thus, many BaseNodes type nodes can be chained together for desired transformations"
```
Comment (andrewkho) [inline examples/nodes/torchata_nodes_basics.ipynb:338]: This should probably be markdown, maybe use this in the cell below instead of my suggestion

File: examples/nodes/torchata_nodes_basics.ipynb [hunk 11]
```diff
@@ -0,0 +1,406 @@
...
+      "[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]\n",
+      "On epoch = 1\n",
+      "[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]\n"
+     ]
+    }
+   ],
+   "source": [
+    "batched_dataset = Batcher(base_node, batch_size = 10)\n",
+    "\n",
+    "# If we try to use this batcher over multiple epochs, we will need to reset it after every epoch\n",
```
Comment (andrewkho) [inline examples/nodes/torchata_nodes_basics.ipynb:279]: This should be markdown, above the cell

File: examples/nodes/torchata_nodes_basics.ipynb [hunk 12]
```diff
@@ -0,0 +1,406 @@
...
+     "output_type": "stream",
+     "text": [
+      "[0, 1, 2, 3]\n",
+      "[4, 5, 6, 7]\n",
+      "[8, 9]\n"
+     ]
+    }
+   ],
+   "source": [
+    "# We can make `drop_last = False` to produce the last batch\n",
```
Comment (andrewkho) [inline examples/nodes/torchata_nodes_basics.ipynb:253]: make this markdown

File: examples/nodes/torchata_nodes_basics.ipynb [hunk 13]
```diff
@@ -0,0 +1,406 @@
...
+     "output_type": "stream",
+     "text": [
+      "[0, 1, 2, 3]\n",
+      "[4, 5, 6, 7]\n"
+     ]
+    }
+   ],
+   "source": [
+    "# A BaseNode can be passed into a Batcher, to get batches of size `batch_size`.\n",
+    "# By default, drop_last is True, meaning if the last batch has a size smaller than the `batch_size`, it is not produced.\n",
```
Comment (andrewkho) [inline examples/nodes/torchata_nodes_basics.ipynb:229]: markdown

File: examples/nodes/torchata_nodes_basics.ipynb [hunk 14]
```diff
@@ -0,0 +1,406 @@
...
+      "25\n",
+      "36\n",
+      "49\n",
+      "64\n",
+      "81\n"
+     ]
+    }
+   ],
+   "source": [
+    "# We can use the Mapper class, to apply a transformation defined using the `map_fn`\n",
```
Comment (andrewkho) [inline examples/nodes/torchata_nodes_basics.ipynb:166]: markdown

**Review by andrewkho — CHANGES_REQUESTED** [review]
Some suggestions, mainly to change all comments to be markdown above the cell.

File: examples/nodes/torchata_nodes_basics.ipynb [hunk 15]
```diff
@@ -0,0 +1,439 @@
...
+      "25\n",
+      "36\n",
+      "49\n",
+      "64\n",
+      "81\n"
+     ]
+    }
+   ],
+   "source": [
+    "# It can also be executed in parallel, using the multi threading/processing approaches, depending on the defined `method`\n",
```
Comment (andrewkho) [inline examples/nodes/torchata_nodes_basics.ipynb:212]: Let's make this markdown too

File: examples/nodes/torchata_nodes_basics.ipynb [hunk 16]
```diff
@@ -0,0 +1,439 @@
...
+    "# Let's look at Loader in the next cell"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d2232823-0927-432b-9171-fb2bf38a7205",
+   "metadata": {},
+   "source": [
+    "#### Loader\n",
+    "As you can see, we get a batch in every epoch, without even needing to reset the loader!!"
```
Comment (andrewkho) [inline examples/nodes/torchata_nodes_basics.ipynb:333]: This would be a good opportunity to showcase SamplerWrapper, and how it will increment epoch automatically, can you maybe add a new cell above here to demonstrate SamplerWrapper, and then how it works together with Loader? We could print the epoch value of eg RandomSampler, and show the different results after shuffling

File: examples/nodes/torchata_nodes_basics.ipynb [hunk 17]
```diff
@@ -0,0 +1,439 @@
...
+    "#### Integrating with torch.data Dataloaders and Samplers"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "97fb43c8-16c0-4c87-bb3a-ff6a0e126764",
+   "metadata": {},
+   "source": [
+    "We can also use torch.data.utils style dataloaders and samplers, and then wrap them into nodes.\n",
+    "Please refer to this [migration guide](https://pytorch.org/data/docs/build/html/migrate_to_nodes_from_utils.html ) to migrate from torch.utils.data to torchdata.nodes\n"
```
Comment (andrewkho) [inline examples/nodes/torchata_nodes_basics.ipynb:105]: Update URL: https://pytorch.org/data/main/migrate_to_nodes_from_utils.html

File: examples/nodes/torchata_nodes_basics.ipynb (hunk 3 above)
Comment (andrewkho) [reply examples/nodes/torchata_nodes_basics.ipynb:131]: Updated: https://pytorch.org/data/main/migrate_to_nodes_from_utils.html

File: examples/nodes/torchata_nodes_basics.ipynb (hunk 16 above)
Comment (ramanishsingh) [reply examples/nodes/torchata_nodes_basics.ipynb:333]: Sure.

File: examples/nodes/torchata_nodes_basics.ipynb [hunk 18]
```diff
@@ -0,0 +1,485 @@
...
+    "#### SamplerWrapper"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "18134e7d-d0e8-4096-a674-23e715fa16a3",
+   "metadata": {},
+   "source": [
+    "As mentioned earlier, we can use `torch.data.utils` samplers using `MapStyleWrapper`.\n",
+    "Alternatively, we can employ the `SamplerWrapper`, which effectively converts a `Sampler` into a `BaseNode`, which can be further used for generating batches. Moreover, `SamplerWrapper` also tracks the number of epochs."
```
Comment (andrewkho) [inline examples/nodes/torchata_nodes_basics.ipynb:379]:
```suggestion
    "Alternatively, we can employ the `SamplerWrapper`, which converts a `Sampler` into a `BaseNode`. `SamplerWrapper` differs from `IterableWrapper` because it will track the number of epochs, and call the sampler's `set_epoch` method if it is implemented."
```

**Review by andrewkho — APPROVED** [review]
One nit in wording and then good to go
